File: server/app/jobs/scheduler.py
from apscheduler.schedulers.background import BackgroundScheduler
from app.jobs.slot_cleanup import cleanup_expired_slot_holds
from app.jobs.medication import dispatch_medication_reminders
from app.jobs.email_retry import process_email_retries
from app.jobs.reminder import dispatch_appointment_reminders
from app.utils.logger import logger

scheduler = BackgroundScheduler()

def start_scheduler():
    scheduler.add_job(cleanup_expired_slot_holds, 'interval', seconds=60, id='slot_cleanup')
    scheduler.add_job(dispatch_medication_reminders, 'interval', minutes=2, id='medication_reminders')
    scheduler.add_job(process_email_retries, 'interval', minutes=2, id='email_retries')
    scheduler.add_job(dispatch_appointment_reminders, 'interval', minutes=15, id='appointment_reminders')
    
    scheduler.start()
    logger.info("All 4 background workers (Slot Holds, Medication, Email Retries, Appointment Reminders) actively scheduled via APScheduler.")

File: server/src/jobs/scheduler.ts
import cron from 'node-cron';
import { cleanupExpiredSlotHolds } from './slotCleanup.job.js';
import { processMedicationReminders } from './medication.job.js';
import { processEmailRetries } from './emailRetry.job.js';
import { processAppointmentReminders } from './reminder.job.js';
import { logger } from '../utils/logger.js';

export function startBackgroundJobs() {
  logger.info('Initializing automated background cron schedulers...');

  // 1. Expired Slot Hold Cleanup: runs every 1 minute
  cron.schedule('* * * * *', async () => {
    await cleanupExpiredSlotHolds();
  });

  // 2. Medication Reminder Dispatcher: runs every 2 minutes
  cron.schedule('*/2 * * * *', async () => {
    await processMedicationReminders();
  });

  // 3. Email Retry Queue Processor: runs every 2 minutes
  cron.schedule('*/2 * * * *', async () => {
    await processEmailRetries();
  });

  // 4. 24h & 2h Appointment Reminder Check: runs every 15 minutes
  cron.schedule('*/15 * * * *', async () => {
    await processAppointmentReminders();
  });

  logger.success('All 4 background workers (Slot Holds, Medication, Email Retries, Appointment Reminders) are actively scheduled.');
}
